src/vdaf/poplar1.rs

```diff
@@ -776,69 +775,68 @@ impl Poplar1AggregationParam {
-        // Thus, setting aside the order of bytes, bits within each byte are ordered with the
-        // [`Msb0`](bitvec::prelude::Msb0) convention, not [`Lsb0`](bitvec::prelude::Msb0). Yet,
-        // the entire integer is aligned to the least significant bit of the last byte, so we
-        // could not use `Msb0` directly without padding adjustments. Instead, we use `Lsb0`
-        // throughout and reverse the bit order of each prefix.
-
-        let mut packed = self
-            .prefixes
-            .iter()
-            .flat_map(|input| input.iter().rev())
-            .collect::<BitVec<u8, Lsb0>>();
-        packed.set_uninitialized(false);
-        let mut packed = packed.into_vec();
-        packed.reverse();
-        bytes.append(&mut packed);
+
+        // Number of prefixes
+        let num_prefixes =
+            u32::try_from(self.prefixes.len()).map_err(|e| CodecError::Other(e.into()))?;
+        num_prefixes.encode(bytes)?;
+
+        // Encoded prefixes
+        for prefix in self.prefixes.iter() {
+            bytes.append(&mut prefix.to_bytes());
+        }
+
         Ok(())
     }
 
     fn encoded_len(&self) -> Option<usize> {
-        let packed_bit_count = (usize::from(self.level) + 1) * self.prefixes.len();
+        let encoded_prefixes_len = (((self.level + 1) as usize + 7) / 8) * self.prefixes.len();
         // 4 bytes for the number of prefixes, 2 bytes for the level, and a variable number of bytes
-        // for the packed prefixes themselves.
-        Some(6 + (packed_bit_count + 7) / 8)
+        // for the encoded prefixes themselves.
+        Some(6 + encoded_prefixes_len)
     }
 }
 
 impl Decode for Poplar1AggregationParam {
     fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        // Level
         let level = u16::decode(bytes)?;
-        let prefix_count =
+
+        // Number of prefixes
+        let num_prefixes =
             usize::try_from(u32::decode(bytes)?).map_err(|e| CodecError::Other(e.into()))?;
 
-        let packed_bit_count = (usize::from(level) + 1) * prefix_count;
-        let mut packed = vec![0u8; (packed_bit_count + 7) / 8];
-        bytes.read_exact(&mut packed)?;
-        if packed_bit_count % 8 != 0 {
-            let unused_bits = packed[0] >> (packed_bit_count % 8);
-            if unused_bits != 0 {
+        // Encoded prefixes
+        let mut prefixes = Vec::with_capacity(num_prefixes);
+        let mut buf = vec![0; ((level + 1) as usize + 7) / 8];
+        let last_byte_mask = match (level + 1) % 8 {
+            0 => 0,
+            num_bits => {
+                let mut mask = 0;
+                for bit_position in 8 - num_bits..8 {
+                    mask |= 1 << bit_position;
+                }
+                mask ^ 255
+            }
+        };
+        for _ in 0..num_prefixes {
+            bytes.read_exact(&mut buf)?;
+            // Ensure there are no trailing bits in the encoded prefix.
```

Comment:
@divergentdave do you think we should specify this case in the draft?

Reply:
I think it would make sense. It can only help catch faulty implementations.

Reply:
https://github.com/cfrg/draft-irtf-cfrg-vdaf/issues/529